File: frontend/src/api/constraints.ts
import { apiFetch, ApiError, readError } from './client'
import type { GridSlot } from '../lib/schedule'

export type ConstraintRecord = {
  id: number
  code: string
  name: string
  description: string
  kind: string
  locked: boolean
  enabled: boolean
  sort_order: number
}

export type ConstraintSummary = {
  hard: number
  soft: number
  soft_enabled: number
  department_rules: number
  current_profile: string | null
  validation_percent: number
}

export type WeightProfile = {
  id: number
  code: string
  name: string
  schedule_stability: number
  student_idle_gaps: number
  room_utilization: number
  lecturer_preferences: number
  daily_balance: number
  building_movement: number
  is_current: boolean
}

export type WeightValues = Omit<WeightProfile, 'id' | 'code' | 'name' | 'is_current'>

export type AvailabilityException = {
  id: number
  starts_on: string
  ends_on: string
  start_period: string | null
  end_period: string | null
  reason: string
  kind: string
}

export type LecturerAvailability = {
  lecturer_id: number
  lecturer_name: string
  submitted: boolean
  coverage_percent: number
  available_slots: number
  preferred_slots: number
  unavailable_slots: number
  slots: GridSlot[]
  exceptions: AvailabilityException[]
}

export type LecturerPreferences = {
  lecturer_id: number
  prefer_morning: boolean
  avoid_friday_afternoon: boolean
  no_early_after_late: boolean
  max_classes_per_day: number
  max_consecutive_hours: number
  min_break_minutes: number
  preferred_days: string[]
}

export type RoomBlock = {
  id: number
  starts_on: string
  ends_on: string
  start_period: string | null
  end_period: string | null
  reason: string
  kind: string
  recurring: boolean
}

export type RoomAvailability = {
  room_id: number
  room_code: string
  slots: GridSlot[]
  blocks: RoomBlock[]
}

async function requestJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(path, init)
  if (!response.ok) {
    throw new ApiError(response.status, await readError(response))
  }
  if (response.status === 204) {
    return undefined as T
  }
  return (await response.json()) as T
}

export function fetchConstraintSummary() {
  return requestJson<ConstraintSummary>('/api/constraints/summary')
}

export function fetchConstraints() {
  return requestJson<ConstraintRecord[]>('/api/constraints')
}

export function patchConstraint(id: number, enabled: boolean) {
  return requestJson<ConstraintRecord>(`/api/constraints/${id}`, {
    method: 'PATCH',
    body: JSON.stringify({ enabled }),
  })
}

export function fetchWeightProfiles() {
  return requestJson<WeightProfile[]>('/api/constraint-weights')
}

export function patchWeightProfile(id: number, body: WeightValues) {
  return requestJson<WeightProfile>(`/api/constraint-weights/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(body),
  })
}

export function activateWeightProfile(id: number) {
  return requestJson<WeightProfile>(`/api/constraint-weights/${id}/activate`, {
    method: 'POST',
  })
}

export function fetchMyAvailability() {
  return requestJson<LecturerAvailability>('/api/me/availability')
}

export function saveMyAvailability(slots: GridSlot[]) {
  return requestJson<LecturerAvailability>('/api/me/availability', {
    method: 'PUT',
    body: JSON.stringify({ slots }),
  })
}

export function createMyException(body: {
  starts_on: string
  ends_on: string
  start_period?: string | null
  end_period?: string | null
  reason: string
  kind?: string
}) {
  return requestJson<AvailabilityException>('/api/me/availability/exceptions', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function deleteMyException(id: number) {
  return requestJson<void>(`/api/me/availability/exceptions/${id}`, {
    method: 'DELETE',
  })
}

export function fetchMyPreferences() {
  return requestJson<LecturerPreferences>('/api/me/preferences')
}

export function saveMyPreferences(body: Omit<LecturerPreferences, 'lecturer_id'>) {
  return requestJson<LecturerPreferences>('/api/me/preferences', {
    method: 'PUT',
    body: JSON.stringify(body),
  })
}

export function fetchRoomAvailability(roomId: number) {
  return requestJson<RoomAvailability>(`/api/rooms/${roomId}/availability`)
}

export function saveRoomAvailability(roomId: number, slots: GridSlot[]) {
  return requestJson<RoomAvailability>(`/api/rooms/${roomId}/availability`, {
    method: 'PUT',
    body: JSON.stringify({ slots }),
  })
}

export function createRoomBlock(
  roomId: number,
  body: {
    starts_on: string
    ends_on: string
    start_period?: string | null
    end_period?: string | null
    reason: string
    kind: string
    recurring?: boolean
  },
) {
  return requestJson<RoomBlock>(`/api/rooms/${roomId}/availability/blocks`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function deleteRoomBlock(roomId: number, blockId: number) {
  return requestJson<void>(`/api/rooms/${roomId}/availability/blocks/${blockId}`, {
    method: 'DELETE',
  })
}

File: backend/app/schemas/constraints.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field


class ConstraintOut(BaseModel):
    id: int
    code: str
    name: str
    description: str
    kind: str
    locked: bool
    enabled: bool
    sort_order: int

    model_config = ConfigDict(from_attributes=True)


class ConstraintEnabledIn(BaseModel):
    enabled: bool


class WeightProfileOut(BaseModel):
    id: int
    code: str
    name: str
    schedule_stability: int
    student_idle_gaps: int
    room_utilization: int
    lecturer_preferences: int
    daily_balance: int
    building_movement: int
    is_current: bool

    model_config = ConfigDict(from_attributes=True)


class WeightProfileIn(BaseModel):
    schedule_stability: int = Field(ge=0, le=10)
    student_idle_gaps: int = Field(ge=0, le=10)
    room_utilization: int = Field(ge=0, le=10)
    lecturer_preferences: int = Field(ge=0, le=10)
    daily_balance: int = Field(ge=0, le=10)
    building_movement: int = Field(ge=0, le=10)


class SlotIn(BaseModel):
    weekday: str
    period: str
    state: str


class SlotOut(BaseModel):
    weekday: str
    period: str
    state: str

    model_config = ConfigDict(from_attributes=True)


class AvailabilityExceptionIn(BaseModel):
    starts_on: date
    ends_on: date
    start_period: str | None = None
    end_period: str | None = None
    reason: str = Field(min_length=1, max_length=255)
    kind: str = "unavailable"


class AvailabilityExceptionOut(BaseModel):
    id: int
    starts_on: date
    ends_on: date
    start_period: str | None
    end_period: str | None
    reason: str
    kind: str

    model_config = ConfigDict(from_attributes=True)


class LecturerAvailabilityOut(BaseModel):
    lecturer_id: int
    lecturer_name: str
    submitted: bool
    coverage_percent: int
    available_slots: int
    preferred_slots: int
    unavailable_slots: int
    slots: list[SlotOut]
    exceptions: list[AvailabilityExceptionOut]


class AvailabilitySlotsIn(BaseModel):
    slots: list[SlotIn]


class LecturerPreferenceIn(BaseModel):
    prefer_morning: bool
    avoid_friday_afternoon: bool
    no_early_after_late: bool
    max_classes_per_day: int = Field(ge=1, le=8)
    max_consecutive_hours: int = Field(ge=1, le=10)
    min_break_minutes: int = Field(ge=0, le=240)
    preferred_days: list[str]


class LecturerPreferenceOut(LecturerPreferenceIn):
    lecturer_id: int

    model_config = ConfigDict(from_attributes=True)


class RoomBlockIn(BaseModel):
    starts_on: date
    ends_on: date
    start_period: str | None = None
    end_period: str | None = None
    reason: str = Field(min_length=1, max_length=255)
    kind: str
    recurring: bool = False


class RoomBlockOut(BaseModel):
    id: int
    starts_on: date
    ends_on: date
    start_period: str | None
    end_period: str | None
    reason: str
    kind: str
    recurring: bool

    model_config = ConfigDict(from_attributes=True)


class RoomAvailabilityOut(BaseModel):
    room_id: int
    room_code: str
    slots: list[SlotOut]
    blocks: list[RoomBlockOut]


class ConstraintSummary(BaseModel):
    hard: int
    soft: int
    soft_enabled: int
    department_rules: int = 0
    current_profile: str | None
    validation_percent: int
